docs/rfcs/xxxx-river.md

```diff
@@ -0,0 +1,400 @@
+language consumed once at load time.
+
+An embedded language is a good fit for Flow:
+
+* It makes it easy for developers to expose APIs which users can interact with
+  or pass around. These APIs can be opaque arbitrary Go types which the user
+  doesn't need to know the detail of, only that it refers to something like a
+  stream of metric samples.
+
+* It is well-suited for continuous evaluation (i.e., the core feature of Flow)
+  so configuration can adapt to a changing environment.
+
+### Why a declarative language? Why not Lua?
+
+The language Flow relies on should have a minimal learning curve. While a
+language like Lua could likely be a decent fit for Flow, imperative languages
+have steeper learning curves compared to declarative languages.
+
+Declarative languages natively map to configuration files, since configuration
+files are used to tell the application the desired state, reducing the learning
+curve for the language and making it easier for users to reason about what the
+final config state should be.
+
+### Why not HCL?
+
+> For some background, it's important to note that HCL can be considered two
+> separate projects: hashicorp/hcl (the language and expression evaluator) and
+> zclconf/go-cty (the value and type system used by HCL).
+
+HCL was the obvious first choice for the Flow prototype: it supports
+expressions, you can expose functions for users to call, and its syntax has a
+small learning curve.
+
+However, I found the [schema-driven processing][] API exposed by HCL to be
+difficult to work with for Flow, requiring a lot of boilerplate. While there is
+a library to interoperate with tagged Go structs, it was insufficient for
+passing around arbitrary Go values, requiring me to [fork][gohcl] both
+github.com/hashicorp/hcl/v2/gohcl and github.com/zclconf/go-cty/cty/gocty to
+reduce boilerplate. This fork contains a non-trivial amount of changes that
+would need to be contributed upstream to be tenable long-term.
+
+Additionally, there is desired functionality that is not supported today in
+HCL/go-cty:
+
+1. A stronger focus on performance and memory usage, changing go-cty to operate
+   around Go values instead of converting Go values to a custom representation.
+2. Ability to disable go-cty's requirement that strings are UTF-8 encoded
+3. Pass around functions as go-cty values (e.g., to allow a clustering
+   component to expose a function to check for ownership of key against a hash
+   ring)
+4. Ability to declare local variables in a scope without needing a `locals`
+   block like as seen in Terraform.
+
+The combination of desired changes across gohcl and go-cty, the fork that was
+already necessary to make it easier to adopt HCL for Flow, and the desire to
+have a stronger interaction with arbitrary Go values led to the decision that a
+new Flow-specific language was warranted.
+
+### Why now?
+
+Grafana Agent Flow is already a dramatic change to the Agent. To avoid users
```

Comment:
Is this new language necessary for an MVP? I was thinking if HCL is enough for the scope of the MVP, should we invest time on design a new language if later users wouldn't be happy to adopt Flow?